File: Python/ExampleTests/common/api_request_string.py
from . import tokens
DNC_STRING = '?DNC?'


class ApiRequestString:
    def __init__(self, method: str = '', page: str = '', parameter: str = '', body: str = ''):
        self.method = method
        self.page = page
        self.parameter = parameter
        self.body = body

    @classmethod
    def from_list(cls, values):
        v = list(values)
        return cls(
            v[0] if len(v) > 0 else '',
            v[1] if len(v) > 1 else '',
            v[2] if len(v) > 2 else '',
            v[3] if len(v) > 3 else ''
        )

    @classmethod
    def from_text(cls, text):
        """Builds from the text form, e.g. Money as '25 USD'."""
        parts = tokens.require(text, 4, 'ApiRequest')
        return cls(
            parts[0],
            parts[1],
            parts[2],
            parts[3]
        )

    def __str__(self):
        return (tokens.token(self.method) + ' ' +
                tokens.token(self.page) + ' ' +
                tokens.token(self.parameter) + ' ' +
                tokens.token(self.body))

    def _key(self):
        return (self.method, self.page, self.parameter, self.body)

    def __eq__(self, other):
        if not isinstance(other, ApiRequestString):
            return NotImplemented
        return all(a == b or a == DNC_STRING or b == DNC_STRING
                   for a, b in zip(self._key(), other._key()))

    def __hash__(self):
        return hash(self._key())

    def __repr__(self):
        return f'ApiRequestString({self})'

File: Swift/ExampleTests/common/ApiRequestString.swift
public struct ApiRequestString: CustomStringConvertible, Equatable {
    public let method: String
    public let page: String
    public let parameter: String
    public let body: String

    public init(method: String, page: String, parameter: String, body: String) {
        self.method = method
        self.page = page
        self.parameter = parameter
        self.body = body
    }

    public init(fromArray v: [String]) {
        self.method = v.count > 0 ? v[0] : ""
        self.page = v.count > 1 ? v[1] : ""
        self.parameter = v.count > 2 ? v[2] : ""
        self.body = v.count > 3 ? v[3] : ""
    }

    /// Builds from the text form, e.g. Money as "25 USD".
    public static func fromText(_ text: String) -> ApiRequestString {
        let parts = Tokens.require(text, 4, "ApiRequest")
        return ApiRequestString(method: parts[0], page: parts[1], parameter: parts[2], body: parts[3])
    }

    public var description: String {
        return Tokens.token(method) + " " + Tokens.token(page) + " " + Tokens.token(parameter) + " " + Tokens.token(body)
    }

    public static let dncString = "?DNC?"

    public static func == (a: ApiRequestString, b: ApiRequestString) -> Bool {
        return (a.method == dncString || b.method == dncString || a.method == b.method)
            && (a.page == dncString || b.page == dncString || a.page == b.page)
            && (a.parameter == dncString || b.parameter == dncString || a.parameter == b.parameter)
            && (a.body == dncString || b.body == dncString || a.body == b.body)
    }
}
